### DISCUSSION TITLE
Process Frigate images with CompreFace

### DISCUSSION THREAD

**Discussion by jakowenko** [exadel-inc/CompreFace#419] [Show and tell]
I love Frigate and have been spending the last few weeks trying to use it for a room presence detection system. I've been using Frigate + CompreFace + Node-Red to achieve this, but my Node-Red flow became pretty complicated and hard to maintain, so I decided to move my logic to a local API, which I containerized.

The API is exposed with a POST method `/recognize` which expects the MQTT frigate/events JSON as the payload.

Images are then processed though [Facebox](https://machinebox.io/docs/facebox) and/or [CompreFace](https://github.com/exadel-inc/CompreFace) for identification.

The output from the API is an array of matching faces, which I then process through Node-Red to update the location for the matching users in Home Assistant.

Subscribing to the  MQTT frigate/events topic directly is also supported and the matches faces are published to a different MQTT topic that can be defined by the user. To see all the available options, checkout the [README](https://github.com/jakowenko/double-take/blob/master/README.md).

```
[
    {
        "id": "1613103956.29759-4g05fi",
        "name": "david",
        "matched": true,
        "confidence": 62.91,
        "attempt": 1,
        "attempts": 1,
        "type": "latest",
        "time": 0.65,
        "camera": "office",
        "room": "Office",
        "totalTime": 0.68,
    }
]
```
Node-Red flow using the REST API
![Screen Shot 2021-02-15 at 5 16 48 PM](https://user-images.githubusercontent.com/1081811/107998417-152c8380-6fb3-11eb-846a-5c650cd701cd.png)

Node-Red flow subscribing to the topic that contains matches after being processed
![Screen Shot 2021-02-25 at 2 31 58 AM](https://user-images.githubusercontent.com/1081811/109119061-27749300-7712-11eb-9db0-18a552b03b1a.png)

![Screen Shot 2021-02-15 at 5 32 37 PM](https://user-images.githubusercontent.com/1081811/107998790-d8ad5780-6fb3-11eb-8c2d-fe0052766571.png)

I was hoping for feedback from the community and wanted to see if others would find something like this useful.

Github: https://github.com/jakowenko/double-take
Docker Hub: https://hub.docker.com/r/jakowenko/double-take

**Comment by pospielov** [comment]
Looks great!
How many frames per second did you achieve? 
Could you share your thoughts about pros and cons of CompreFace and FaceBox?

**Reply by jakowenko** [reply]
I'm currently running 6 RTPS cameras at 2 FPS with Frigate, though I can probably up this to at least 5 FPS.

I've trained both CompreFace and Facebox with the same images. Roughly 20 of myself and 10 of my girlfriend. These were all uncropped and unnormalized images. Would cropping and normalizing improve the results?

CompreFace seemed to detect myself quicker when I was the only person in the house. Usually > 80% confidence level. But when my girlfriend was over, CompreFace almost every time mixed her up for myself. Even with similarity levels of > 70%.

Facebox does seem to be a little slower, but I haven't actually looked at the raw numbers yet. Though it did seem to better identify me vs my girlfriend.

I'm going to update my container to better test the two side by side. Maybe create an endpoint where you can stand in front of a camera and it would output the results from each service with the times it took to find a match.

If you have any ideas you'd like to see me incorporate to help test them, let me know.

**Reply by pospielov** [reply]
These are quite strange results. I see it the first time when CompreFace mixes people when there are only two people in the face collection.
When I did my experiments, I uploaded about 50 000 different subjects to one face collection and about 500 images of myself in another. Then I took another photo of myself and uploaded it to both collections. The similarity result is in the picture:
![image](https://user-images.githubusercontent.com/3736126/109384395-16d04280-78f5-11eb-8053-adc1ea5c49ca.png)
The red series is similarities to me, and the blue is similarities to other people
As you can see the probability of mixing up is very low.
Could you send the "recognize" REST with your girlfriend with "prediction_count=50" query parameter and share the response with me? 

Also, CompreFace crops and normalize images, so I believe it won't improve accuracy if you do it yourself. What could improve - upload examples of the face with different angles of rotation, so it will have the example for every situation

**Reply by jakowenko** [reply]
I haven't had a chance to try that prediction_count flag yet on the issues I had when my girlfriend was over, but here's a strange false positive I had today.

Frigate detected my cat as a person, but CompreFace then detected my couch pillow as me with a pretty high similarity.

![Screen Shot 2021-03-04 at 1 17 32 AM](https://user-images.githubusercontent.com/1081811/109919881-6c0aac00-7c87-11eb-91c2-cb40d48d182f.png)

**Reply by pospielov** [reply]
Could you send me the original picture? 
I tried to cut it from the screenshot and CompreFace didn't find any faces:
![2021-03-04_22-05](https://user-images.githubusercontent.com/3736126/110023864-5607f480-7d36-11eb-9efb-b548ae1b6d93.png)
Also, look at the probability score - it's 0.76237, this is low for detection. In my experience even faces on bad quality photos have a probability of more than 0.9.
I looked at the code, looks like the default threshold is 0.65 now, probably need to change it in the next release :)
For now, you can use `det_prob_threshold` query param to change its default value.
The problem with the similarity still exists, don't understand how it could be so high...

**Reply by jakowenko** [reply]
Here's the raw photo. Thanks for the tip with the `det_prob_threshold`. I was looking at that last night, I'll try to use that with a value of .9 moving forward to see how the results look.

![1614819751 960475-r3gyaa-latest](https://user-images.githubusercontent.com/1081811/110027034-78cae680-7cff-11eb-9b96-61fc59c766a1.jpg)

**Reply by pospielov** [reply]
So I checked your image and I can confirm, it detects pillow as a face :)
I made an experiment: I uploaded to a face collection 300 images of myself with different quality and named them as different subjects.
Then I uploaded this picture.
4 images of 300 crossed the threshold of 0.7 (I believe this is a threshold for face recognition), even for two of the images the similarity was more than 0.9 which is extremely high. All these 4 photos have awful quality. 
Then I uploaded the image of you from your first post. As expected the similarity wasn't more than 0.52.

Then I made another experiment. I took 50000 photos of people from a public dataset and uploaded them to face collection.
And uploaded the photo with the pillow. There were 37 results with a similarity of more than 0.9. I checked the original photos and all of them have bad quality, mostly very dark.
Then I took the first photo that is "similar" to the pillow and uploaded it again to this 50000 photos collection. It returned a similar result with the pillow - 13 of 20 most "similar" faces were the same.
From this experiment, I can tell that the face recognition system works badly with faces with bad quality, especially for dark faces. 

So what I would recommend you - upload faces with good quality to face collection. I think it could work better even if the test images are of bad quality.

**Reply by jakowenko** [reply]
Thanks for looking into this and reporting back your findings. I'll make sure to feed CompreFace better pictures of myself too. I've completely dropped using Facebox because CompreFace was working so well. Plus Facebox has a 100 image training limit, which is a bummer.

I do from time to time still get false positives between myself, my sister, and my girlfriend, but I'm hoping as I train the model with more images of each of us, then the results will improve.

**Comment by jakowenko** [comment]
I've made quite a few UI updates over the past few weeks and the root of the API at `http://YOUR-IP:PORT` will now serve a frontend that shows all matches and misses from the detectors. You can either retrain the models with the images or delete them. More updates to follow! Would love some feedback if anybody tries it out.

<img src="https://community-assets.home-assistant.io/original/3X/8/3/83675c23e56fa5750b6a7faef8a77a00d33ac96a.png">

**Reply by pospielov** [reply]
I see from your screenshots that some of the results were made by CompreFace and some from Deepstack.
Do you use some ensemble of solutions to predict? 
Like if both CompreFace and Deepstack predict that this is David, then you accept the result and if not, then do not accept?

**Reply by jakowenko** [reply]
I use both CompreFace and DeepStack currently and take the best result between them for each event that is processed.

Interesting idea though to maybe use the combination of both of them before confirming it's a positive match.

**Comment by jakowenko** [comment]
I pushed quite a few updates and small performance improvements last night, including support for CompreFace v0.5.0. If you are using multiple detectors, then the same image is passed on each processing loop. This allows me to display a single image on the UI with the results per detector.

<img src="https://user-images.githubusercontent.com/1081811/116505792-cc822800-a889-11eb-91f3-f6a192767aa8.png" width="100%">

I've also added some more [documentation](https://github.com/jakowenko/double-take/blob/master/docs/home-assistant-node-red.md) on how I setup my Node-Red and Home Assistant notifications.

More notes on what's included in this release can be found [here](https://github.com/jakowenko/double-take/releases/tag/v0.3.2).

**Comment by jakowenko** [comment]
More updates 🚀

CompreFace has definitely been the best performer so far, so thank you again for the amazing work!

- I've migrated the ENV variables to a more readable [`config.yml`](https://github.com/jakowenko/double-take#configuration) file.
  Here's a small sample:
  ```yaml
  server:
    port: 3000

  mqtt:
    host: 192.168.1.1

  home_assistant:
    url: http://192.168.1.1:8123
    token: xxx.xxx-xxx

  confidence:
    match: 60
    unknown: 40

  save:
    matches: true
    unknown: true

  frigate:
    url: http://192.168.1.1:4000

  detectors:
    compreface:
      url: http://192.168.1.1:8000
      key: xxx-xxx-xxx-xxx-xxx
    deepstack:
      url: http://192.168.1.1:8001
  ```
- Added Home Assistant support. Entities for people and cameras will automatically be created

Any feedback or feature suggestions would be appreciated.

**Reply by pospielov** [reply]
Hi, sounds great!

I visited your repository to understand it better and probably give some feedback.
And here is the problem - I haven't used Frigate, Home Assistant, and Node-Red.
For me, it doesn't clear what exactly your solution does. 
"Unified UI and API for processing and training images for facial recognition." - sounds exciting, but then I read about your use cases and don't get it. I mean I assume I can't just use it to train CompreFace face recognition service, without Frigate or Home Assistant, right?
So what I suggest:
1. Be more clear in the description - for whom this solution? E.g. Frigate users that want to add face recognition.
2. The requirements, e.g. Frigate version XX
3. I would be really happy if there was a video demo with installation and first test (I think we need it in CompreFace too :) )

**Reply by jakowenko** [reply]
This is great feedback, thank you!

I agree, I need to better explain some of the use cases for people who don't use any of those third party applications.

**Comment by jakowenko** [comment]
`1.0.0` is officially out with support for web sockets, pagination, updated filtering, and more!

The [configuration](https://github.com/jakowenko/double-take#configuration) section of the README is also updated with better examples.

Thank you to everyone who has used and help test Double Take over the past few months and for all the awesome feature suggestions. Keep them coming!

**Reply by pospielov** [reply]
Wow, congratulations! 
BTW your documentation is much better now! I really like that you published Postman documentation. We also will publish it very soon :)

**Reply by jakowenko** [reply]
Thank you! It's been a journey over the last 6 months haha. I'm telling everybody I know about CompreFace, keep up the amazing work! It's been fun to update my project as you add more features.

**Comment by jakowenko** [comment]
## [v1.4.0](https://github.com/jakowenko/double-take)

### Bug Fixes

* **api:** validate mime type and result of database.get.fileByFilename
* **api:** catch facebox error before normalizing data
* **api:** delete orphaned db records when deleting training folder / files
* **ui:** fix multiselect dropdowns when open and scrolling
* **ui:** theme wouldn't load if setting to same theme

### Features

* support for `ARMv7` architecture
* **api:** `api/latest/<name>.jpg`
* **api:** use camera name when using custom mqtt topic
* **ui:** enable/disable sockets on matches page

**Reply by pospielov** [reply]
Congrats on a new release!
Do you have any ideas on how we can improve CompreFace?

**Reply by jakowenko** [reply]
CompreFace has been great, I really appreciate what you've built so far. Do you have a roadmap so I can get an idea of what you have planned?

I think a cool request I've seen from time to time, not sure if it's feasible, would be to have an ability to train false positives or tell CompreFace that it got the recognition wrong.

**Reply by pospielov** [reply]
A defined roadmap is quite difficult as our progress depends on what contributors do we have. 
But I understand how it's important, I think I'll add something to the readme.
The main ideas to improve CompreFace (the order may be random, depending on the contribution skills):
1. Add UI to manage face collection
2. Create or find a new face recognition model, based on a dataset with better distribution (gender, nationalities, etc.)
3. Improve the performance. Now we just reuse the most popular face recognition libraries, but they are not optimized for performance.
4. Add support of ARM devices and TPU accelerators, like Google Coral
5. Integrate with the most popular solutions like Home Assistant, BlueIris, Shinobi.
6. Add much more examples and videos

>  train false positives or tell CompreFace that it got the recognition wrong.

This is technically impossible, we take the face recognition model and use it for all recognitions. 
You still can retrain this model by yourself and then build CompreFace using your model. Probably we can do something to simplify this process.

**Reply by alexdelprete** [reply]
> This is technically impossible

Pospielov, can you explain for me why it's technically impossible? I'm an old-school dinosaur, and in the old days of bayesian filters to detect email spam, negative training (telling the filter it detected wrongly a legit email) was critical in order to improve and refine its efficiency. Many Double Take users (I was one of the first asking David about this) are asking why we can't tell the engine that it was a false/wrong detection thinking it can "learn" from its mistakes. 

I guess we're not yet in AI with CompreFace, right? :)

Thanks for your amazing work and most of all, your support to the community, it's great how you support the solution and are open to discussing it to improve it, thank you for this, it's highly appreciate it.

**Reply by pospielov** [reply]
https://medium.com/@ageitgey/machine-learning-is-fun-part-4-modern-face-recognition-with-deep-learning-c3cffc121d78
Here is a good article to understand how face recognition works. Of course, there were lots of improvements from 2016, but the main idea remains the same:
1. You take a huge dataset and train your huge convolutional neural network. It could take days even on a powerful GPU.
2. Then you remove fully connected layers on this CNN, so there is a vector of 512(or 128) numbers at the end of the neural network instead of the answer who is this person
3. Then you receive this vector for each saved face (I believe this is called training in double-take, but in CompreFace we just collect vectors, no ML here, so no training)
4. When you need to recognize a face, you calculate the vector for an unknown face, then find the euclidian distance (linear algebra, again without ML) between the unknown face and all other faces. This distance answers the question of whether two faces look similar. 
So, in the result application, there is no ML or training, if the CNN model is bad, we can't improve it. 
Of course, technically it's possible. But the training is not so easy process, there are lots of things to consider - how many layers to train, what training coefficients and iterations. And it depends on the goals and number of examples. I believe if you know all this stuff, you can make it with the script(you can find them in FaceNet or InsightFace libraries) and then just put it into CompreFace. I already thought if we need to add the functionality of uploading a new model from UI. But I believe it won't be a popular feature.

**Reply by alexdelprete** [reply]
> 4\. When you need to recognize a face, you calculate the vector for an unknown face, then find the euclidian distance (linear algebra, again without ML) between the unknown face and all other faces. This distance answers the question of whether two faces look similar.

So basically the vector represents a physical measurement of specific parts of a face, a sort of digital print of the face, and when you scan a new face and take its vector, you search in the neural network for a close enough vector?

I will read the article, thank you very much.

**Reply by pospielov** [reply]
I use a neural network to get a vector. Then just calculate euclidian distance to find the similarity

**Comment by jakowenko** [comment]
I've recently published version `1.12.0` with includes support for AWS Rekognition and I wanted to just say that CompreFace is just as good if not better. Hats off to you guys for providing such an amazing open source project. I wanted to get Rekognition added just so I could test it side by side the other detectors. We'll see over time how it does, but I prefer the price of CompreFace 🍺 


## [1.12.0](https://github.com/jakowenko/double-take/compare/v1.11.0...v1.12.0) (2022-06-13)


### Features

* **api:** opencv preprocess face check ([ed30ad1](https://github.com/jakowenko/double-take/commit/ed30ad10dfe9f6d104619fa134e2d54698b1c1ff))
* aws rekognition support ([7904852](https://github.com/jakowenko/double-take/commit/79048520fd66cb9c4144db11488558b82c8ada19))
* **detectors:** process images from specific cameras ([5d39d0c](https://github.com/jakowenko/double-take/commit/5d39d0c2c18d0851eccd6a8f698638906b0c36b0))
* **frigate:** sort sub labels alphabetically [#217](https://github.com/jakowenko/double-take/issues/217) ([82d8736](https://github.com/jakowenko/double-take/commit/82d8736825b428b2b74385975961233d167db294))
* **frigate:** stop_on_match config option to break process loop ([4b98990](https://github.com/jakowenko/double-take/commit/4b9899077c74da601fa5d497c929c264e8c03fa6))
* **opencv:** adjust classifier settings via config ([2e6c512](https://github.com/jakowenko/double-take/commit/2e6c5129b445b2eca8b884af8d95fc2ceb19733d))
* **ui:** show config errors ([ddcaf89](https://github.com/jakowenko/double-take/commit/ddcaf897ba7f35ac65a6a86a71cbd56a1a9a2a46))
* **ui:** upload images to process with detectors ([f774406](https://github.com/jakowenko/double-take/commit/f774406b540df8ff0eec8b247fee4bd3e02579e1))


### Bug Fixes

* **api:** sleep if image hasn’t changed during retry loop ([78e9808](https://github.com/jakowenko/double-take/commit/78e98082d7e0e77d5c23465cc629b38d9b178a27))
* better handling of update check ([6de1cf1](https://github.com/jakowenko/double-take/commit/6de1cf1dca9e7f244a4242c3c540ef464824bbc3))

**Reply by pospielov** [reply]
I am happy that your project continues to develop. 
Just curious, as I understand, your solution sends requests for recognition only if Frigate founds the face, right? So it shouldn't be so many requests per month. Do you know how many recognitions per month you do? 

Also, `api: opencv preprocess face check ([ed30ad1](https://github.com/jakowenko/double-take/commit/ed30ad10dfe9f6d104619fa134e2d54698b1c1ff))` - did you add another face detection? But why?

**Reply by jakowenko** [reply]
Thank you @pospielov!

Double Take is triggered when Frigate detects a person, so there is a chance a face might not be in the frame when passing the image to the detector. And because DT pulls frames to process, I thought it would be a nice touch to run it through OpenCV if the user wants.

I have 6 cameras setup within my house, so running Rtekognition on all of those would probably start to add up haha. I managed to run up around 5,000 requests with 2 cameras over the course of a few days.
